ledger/participant-integration-api/src/main/scala/platform/store/ActiveLedgerStateManager.scala

```diff
@@ -69,146 +100,174 @@ private[platform] class ActiveLedgerStateManager[ALS <: ActiveLedgerState[ALS]](
-                ats.copy(
-                  errs = contractCheck(ne.targetCoid).fold(errs)(errs + _),
-                  acc = Some(if (ne.consuming) {
-                    acc.removeContract(ne.targetCoid)
-                  } else {
-                    acc
-                  }),
-                  parties = parties.union(nodeParties),
-                  archivedIds = if (ne.consuming) archivedIds + ne.targetCoid else archivedIds,
-                )
-              case nlkup: N.NodeLookupByKey[ContractId] =>
-                // Check that the stored lookup result matches the current result
-                val key = nlkup.key.key.ensureNoCid.fold(
-                  coid =>
-                    throw new IllegalStateException(s"Contract ID $coid found in contract key"),
-                  identity,
+                val newState = acc.copy(
+                  errs = contractCheck(als, ne.targetCoid).fold(acc.errs)(acc.errs + _),
+                  currentState = acc.currentState.copy(
+                    als = Some(if (ne.consuming) {
+                      als.removeContract(ne.targetCoid)
+                    } else {
+                      als
+                    })
+                  ),
+                  parties = acc.parties.union(nodeParties),
                 )
-                val gk = GlobalKey(nlkup.templateId, key)
-                val nodeParties = nlkup.key.maintainers
-
-                if (actAs.nonEmpty) {
-                  // If the submitter is known, look up the contract
-                  // Submitters being know means the transaction was submitted on this participant.
-                  val currentResult = acc.lookupContractByKey(gk)
-                  if (currentResult == nlkup.result) {
-                    ats.copy(
-                      parties = parties.union(nodeParties)
-                    )
-                  } else {
-                    ats.copy(
-                      errs = errs + Inconsistent(
-                        s"Contract key lookup with different results: expected [${nlkup.result}], actual [${currentResult}]"
-                      )
-                    )
-                  }
-                } else {
-                  // Otherwise, trust that the lookup was valid.
-                  // The submitter being unknown means the transaction was submitted on a different participant,
-                  // and (A) this participant may not know the authoritative answer to whether the key exists and
-                  // (B) this code is called from a Indexer and not from the sandbox ledger.
-                  ats.copy(
-                    parties = parties.union(nodeParties)
-                  )
-                }
+                (newState, true)
             }
-        }
-
-    val divulgedContractIds = divulgence -- st.archivedIds
+          },
+          rollbackBegin = (acc, _, _) => {
+            (acc.copy(rollbackStates = acc.currentState +: acc.rollbackStates), true)
```

Comment:
This saving of state checkpoints works if `ALS` is a value (like used in the in-memory sandbox classic [here](https://github.com/digital-asset/daml/blob/main/ledger/sandbox-classic/src/main/scala/platform/sandbox/stores/InMemoryActiveLedgerState.scala#L21)).

If however it's an interface to some underlying mutable storage (like used in some ancient migration [here](https://github.com/digital-asset/daml/blob/main/ledger/participant-integration-api/src/main/scala/db/migration/postgres/V2_1__Rebuild_Acs.scala#L303)), then you will actually keep modifying the "final" ACS after entering rollback nodes.

This is ok because the above mentioned migration will never run on transactions with rollback nodes.

I assume we still need to check the activeness and ledger time of contracts mentioned within rollback node subtrees? Otherwise we could simply skip the rollback nodes entirely.

Reply:
> I assume we still need to check the activeness and ledger time of contracts mentioned within rollback node subtrees? Otherwise we could simply skip the rollback nodes entirely.

Yes everything under a rollback node still needs to be valid but the writes resulting from it (i.e., creates and archives) will be rolled back so we do need to be rolled back.

You do make a valid point. Do you have an idea for how to make this better? Should I add a method to the interface to clone an ALS which can be trivially implemented for the in-memory version and can simply throw in the ancient migration (and hopefully we never will use it anywhere new).

Reply:
:+1: I don't have a better idea, not without a major rewrite of this class.